fs/dafsserver/dafsserver.cpp

```diff
@@ -2915,93 +2915,95 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
             }
         }
         bool isRowServiceClient() const { return calledByRowService; }
-        bool notifySelected(ISocket *sock,unsigned selected)
+        bool notifySelected(ISocket *sock, unsigned selected)
         {
             if (TF_TRACE_FULL)
                 PROGLOG("notifySelected(%p)",this);
             if (sock!=socket)
                 WARNLOG("notifySelected - invalid socket passed");
-            size32_t avail = (size32_t)socket->avail_read();
-            if (avail)
-                touch();
-            else if (left)
-            {
-                WARNLOG("notifySelected: Closing mid packet, %d remaining", left);
-                msg.clear();
-                parent->notify(this, msg); // notifying of graceful close
-                return false;
-            }
-            if (left==0)
+            touch();
+            try
             {
-                try
-                {
-                    left = avail?receiveDaFsBufferSize(socket):0;
-                }
-                catch (IException *e)
-                {
-                    EXCLOG(e,"notifySelected(1)");
-                    e->Release();
-                    left = 0;
-                }
-                if (left)
+                if (!gotSize)
                 {
-                    // TLS TODO: avail_read() may not return accurate amount of pending bytes
-                    avail = (size32_t)socket->avail_read();
-                    try
-                    {
-                        msg.ensureCapacity(left);
-                    }
-                    catch (IException *e)
+                    // left represents amount we have read of leading size32_t (normally expect to be read in 1 go)
+                    if (0 == msg.length()) // 1st time
+                        msgWritePtr = (byte *)msg.reserveTruncate(sizeof(size32_t));
+                    size32_t szRead;
+                    sock->read(msgWritePtr, 1, sizeof(size32_t)-left, szRead);
```

Comment:
why not just make the min_size sizeof(size32_t) here ?

Reply:
because don't want to block, which readtms would do (loop internally), waiting for that min_size.
We have been notified there's something on the wire, but we don't know how much, could be < sizeof(size32_t)

Reply:
Yes, can we make it use readtms() with a small timeout here and below.